File: server/chatbotService.ts
import OpenAI from 'openai';
import type { Storage } from './storage';

function getOpenAIClient(): OpenAI | null {
  const baaSignedEnv = process.env.OPENAI_BAA_SIGNED;
  const isBaaSigned = baaSignedEnv === 'true' || baaSignedEnv === '1';
  
  if (!isBaaSigned || !process.env.OPENAI_API_KEY) {
    return null;
  }
  
  return new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
}

interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

interface ChatbotContext {
  doctorId: string;
  clinicName?: string;
  sessionId: string;
  messages: ChatMessage[];
}

class ChatbotService {
  private storage: Storage;
  private sessions: Map<string, ChatbotContext> = new Map();

  constructor(storage: Storage) {
    this.storage = storage;
  }

  async initializeSession(doctorId: string, sessionId: string): Promise<ChatbotContext> {
    const doctor = await this.storage.getUser(doctorId);
    const clinicName = `Dr. ${doctor?.firstName} ${doctor?.lastName}'s Office`;

    const systemPrompt = `You are Assistant Lysa, a helpful and professional AI receptionist for ${clinicName}. Your role is to:
1. Help patients schedule appointments
2. Answer common questions about services, hours, and location
3. Collect patient information (name, contact, reason for visit)
4. Provide general health information (but always recommend consulting the doctor for specific medical advice)
5. Be empathetic, professional, and HIPAA-compliant

NEVER:
- Provide specific medical diagnoses or treatment recommendations
- Share patient information with unauthorized individuals
- Make promises about medical outcomes

Office hours: Monday-Friday, 9 AM - 5 PM
Location: 123 Medical Plaza, Suite 100

If a patient wants to schedule an appointment, collect: name, phone, email, preferred date/time, and reason for visit.`;

    const context: ChatbotContext = {
      doctorId,
      clinicName,
      sessionId,
      messages: [
        {
          role: 'system',
          content: systemPrompt,
        },
        {
          role: 'assistant',
          content: `Hello! I'm Assistant Lysa from ${clinicName}. How can I help you today?`,
        },
      ],
    };

    this.sessions.set(sessionId, context);
    return context;
  }

  async sendMessage(
    sessionId: string,
    message: string,
    doctorId?: string
  ): Promise<{ response: string; intent?: string; extractedData?: any }> {
    try {
      let context = this.sessions.get(sessionId);

      if (!context) {
        if (!doctorId) {
          throw new Error('Doctor ID required for new session');
        }
        context = await this.initializeSession(doctorId, sessionId);
      }

      context.messages.push({
        role: 'user',
        content: message,
      });

      const openai = getOpenAIClient();

      if (!openai) {
        const fallbackResponse = this.generateFallbackResponse(message);
        context.messages.push({
          role: 'assistant',
          content: fallbackResponse,
        });
        return { response: fallbackResponse };
      }

      const completion = await openai.chat.completions.create({
        model: 'gpt-4o',
        messages: context.messages as any,
        temperature: 0.7,
        max_tokens: 500,
      });

      const response = completion.choices[0]?.message?.content || 'I apologize, but I encountered an error. Please try again.';

      context.messages.push({
        role: 'assistant',
        content: response,
      });

      const intent = await this.extractIntent(message);
      const extractedData = await this.extractPatientData(context.messages);

      this.sessions.set(sessionId, context);

      return {
        response,
        intent,
        extractedData,
      };
    } catch (error) {
      console.error('Chatbot error:', error);
      return {
        response: 'I apologize, but I encountered an error. Please try again or call our office directly.',
      };
    }
  }

  private generateFallbackResponse(message: string): string {
    const lowerMessage = message.toLowerCase();

    if (lowerMessage.includes('appointment') || lowerMessage.includes('schedule')) {
      return "I'd be happy to help you schedule an appointment! To get started, I'll need your name, phone number, email, and the reason for your visit. What works best for you?";
    }

    if (lowerMessage.includes('hours') || lowerMessage.includes('open')) {
      return "Our office hours are Monday through Friday, 9 AM to 5 PM. Is there anything else I can help you with?";
    }

    if (lowerMessage.includes('location') || lowerMessage.includes('address')) {
      return "We're located at 123 Medical Plaza, Suite 100. Is there anything else you'd like to know?";
    }

    return "Thank you for your message. How can I assist you today? I can help with scheduling appointments, answering questions about our services, or providing general information about our office.";
  }

  private async extractIntent(message: string): Promise<string> {
    const lowerMessage = message.toLowerCase();

    if (lowerMessage.includes('appointment') || lowerMessage.includes('schedule')) {
      return 'schedule_appointment';
    }
    if (lowerMessage.includes('cancel')) {
      return 'cancel_appointment';
    }
    if (lowerMessage.includes('reschedule')) {
      return 'reschedule_appointment';
    }
    if (lowerMessage.includes('hours') || lowerMessage.includes('when open')) {
      return 'office_hours';
    }
    if (lowerMessage.includes('location') || lowerMessage.includes('address')) {
      return 'office_location';
    }

    return 'general_inquiry';
  }

  private async extractPatientData(messages: ChatMessage[]): Promise<any> {
    const userMessages = messages
      .filter((m) => m.role === 'user')
      .map((m) => m.content)
      .join(' ');

    const emailRegex = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/;
    const phoneRegex = /\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/;
    const nameRegex = /my name is ([A-Z][a-z]+ [A-Z][a-z]+)/i;

    const email = userMessages.match(emailRegex)?.[0];
    const phone = userMessages.match(phoneRegex)?.[0];
    const name = userMessages.match(nameRegex)?.[1];

    return {
      name: name || null,
      email: email || null,
      phone: phone || null,
    };
  }

  async getChatHistory(sessionId: string): Promise<ChatMessage[]> {
    const context = this.sessions.get(sessionId);
    if (!context) {
      return [];
    }
    return context.messages.filter((m) => m.role !== 'system');
  }

  async endSession(sessionId: string): Promise<void> {
    this.sessions.delete(sessionId);
  }
}

export let chatbotService: ChatbotService;

export function initChatbotService(storage: Storage) {
  chatbotService = new ChatbotService(storage);
}

File: app/services/chatbot_service.py
from typing import Dict, List
from sqlalchemy.orm import Session
from app.models.chatbot import ChatbotConversation
from app.config import settings, check_openai_baa_compliance
import openai


class ChatbotService:
    def __init__(self, db: Session):
        self.db = db
        self.openai_enabled = check_openai_baa_compliance()
        if self.openai_enabled and settings.OPENAI_API_KEY:
            openai.api_key = settings.OPENAI_API_KEY
    
    def get_conversation_history(self, doctor_id: str, session_id: str, limit: int = 10) -> List[Dict]:
        conversations = self.db.query(ChatbotConversation).filter(
            ChatbotConversation.doctor_id == doctor_id,
            ChatbotConversation.session_id == session_id
        ).order_by(ChatbotConversation.created_at.desc()).limit(limit).all()
        
        return [
            {
                "user_message": conv.user_message,
                "bot_response": conv.bot_response,
                "created_at": conv.created_at.isoformat()
            }
            for conv in reversed(conversations)
        ]
    
    async def chat(self, doctor_id: str, session_id: str, message: str) -> Dict:
        if not self.openai_enabled:
            fallback_response = self._get_fallback_response(message)
            conversation = ChatbotConversation(
                doctor_id=doctor_id,
                session_id=session_id,
                user_message=message,
                bot_response=fallback_response
            )
            self.db.add(conversation)
            self.db.commit()
            
            return {
                "response": fallback_response,
                "using_ai": False
            }
        
        history = self.get_conversation_history(doctor_id, session_id)
        
        messages = [
            {"role": "system", "content": "You are a helpful medical clinic assistant. Provide professional, concise responses to help doctors manage their practice."}
        ]
        
        for conv in history[-5:]:
            messages.append({"role": "user", "content": conv["user_message"]})
            messages.append({"role": "assistant", "content": conv["bot_response"]})
        
        messages.append({"role": "user", "content": message})
        
        try:
            response = await openai.ChatCompletion.acreate(
                model="gpt-4o",
                messages=messages,
                max_tokens=500,
                temperature=0.7
            )
            
            bot_response = response.choices[0].message.content
        except Exception as e:
            print(f"OpenAI API error: {e}")
            bot_response = self._get_fallback_response(message)
        
        conversation = ChatbotConversation(
            doctor_id=doctor_id,
            session_id=session_id,
            user_message=message,
            bot_response=bot_response
        )
        self.db.add(conversation)
        self.db.commit()
        
        return {
            "response": bot_response,
            "using_ai": self.openai_enabled
        }
    
    def _get_fallback_response(self, message: str) -> str:
        message_lower = message.lower()
        
        if "appointment" in message_lower or "schedule" in message_lower:
            return "I can help you with appointments. Please check the appointments section to view, create, or modify appointments."
        elif "patient" in message_lower:
            return "For patient information, please use the patient management section of the dashboard."
        elif "reminder" in message_lower:
            return "Appointment reminders are automatically sent 24 hours before scheduled appointments via SMS and email."
        else:
            return "I'm here to help! You can ask me about appointments, patient management, reminders, or general clinic operations."
